### PR TITLE
[Syncer] Fix flaky node sync

### LINKED ISSUE #510: Non-validator node slow to write blocks when connected to other non-validator peers
### Description

There seems to be a very noticeable delay in writing blocks when making `eth_getBlockByNumber(latest)` calls to a non-validator node when connected to other non-validator peers.

We are using an elastic load balancer to manage an RPC cluster which is comprised of 4 non-validating nodes. When directly hitting one of these 4 nodes on cluster scale-up (from 1 running node to a total of 4), we can observe that blocks are enqueued up to the latest block in our chain (looking at `polygon.consensus.syncer: enqueue block` entries) - however, the node only seems to periodically write these blocks. As a result, when making a call to `eth_getBlockByNumber(latest)`, we often see an outdated block.

Because of the above behaviour, when running multiple non-validator nodes within a LB cluster, using a script to spam our RPC endpoint with `eth_getBlockByNumber(latest)` calls will return jsonrpc responses that don't always align with the latest block in our chain.

### Your environment

* OS and version `Ubuntu 20`
* version of the Polygon Edge `8dc4251`
* branch that causes this issue `develop`

Our testnet is made up of 4 validating nodes (+1 non-validating node for making periodic backups), as well as our RPC cluster which is running 4 non-validating nodes.

### Steps to reproduce

1. Spin up 1 non-validator node and have it fully synced to the rest of the peers in the chain.

2. Run script below where RPC_URL is the url of node-1. Observe that the block number increments as expected.
```
async function main() {
  const provider = new providers.JsonRpcProvider(RPC_URL);

  setInterval(async () => {
    const res = await provider.getBlock("latest");
    console.log(res)
  }, 400);
}
```

3. Spin up 3 more non-validator nodes. Observe that while these nodes syncing or even fully-synced, responses to node-1 using the script above would sometimes show the block number to stop incrementing -- see log below.

### Expected behaviour

I wouldn't expect a noticeable delay between enqueuing blocks vs actually writing them. 

### Actual behaviour

I notice an unexpected lag when making jsonrpc requests indicating an outdated block number (even so the logs show that the node has synced up to its peers).

### Logs

Here an initial `eth_getBlockByNumber(latest)` is called which returns `3098003`.  One can see that the node is currently synced to block `3098007`. A subsequent call to `eth_getBlockByNumber(latest)` still returns `3098003`.

```
Apr 20 18:28:25 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:25.946Z [DEBUG] polygon.dispatcher: request: method=eth_getBlockByNumber id=170
Apr 20 18:28:25 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:25.946Z [DEBUG] polygon.jsonrpc: handle: response="{\"jsonrpc\":\"2.0\",\"id\":170,\"result\":{\"parentHash\":\"0xc27f5204d57432915729404a3446d91c439ec3156b509403f576cefd44874e19\",\"sha3Uncles\":\"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347\",\"miner\":\"0x0000000000000000000000000000000000000000\",\"stateRoot\":\"0x037444c8f8dc64d3aa0e84979a4452e72908b34cc17dec99c3b7020b95ac92c4\",\"transactionsRoot\":\"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421\",\"receiptsRoot\":\"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421\",\"logsBloom\":\"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\",\"difficulty\":\"0x2f4593\",\"totalDifficulty\":\"0x2f4593\",\"size\":\"0x3a1\",\"number\":\"0x2f4593\",\"gasLimit\":\"0x3938700\",\"gasUsed\":\"0x0\",\"timestamp\":\"0x626050c2\",\"extraData\":\"0x0000000000000000000000000000000000000000000000000000000000000000f90179f86994a3a585c6d59cce6aae7035e8df48b3327cc8be54947eff16a34e3433182d636488bc97919b10283f37943e64f88c6c7a1310236b242180c0ba1409d10f4d949cde4f11ea57163023a9a8a5df499b27813639eb946c230a5f8dd3413aa04e0860dd225f887b36f101b8416f7b503eb78cc23f7509e2ce456726470a58feb3cd7dd38e4e2c12e57cebf710064d3e7f49eb3ac0e92a4f579d3ecc2461986cc5c95025b455939079da72008501f8c9b841e7ff6a9203233b644d863ce2d86dfdefa77ad738d18d22a2cb25167c18384e4358e01a4f1d5f6d9e4b396e65a3f8b47a43444acf6b42fdc79c7cd786f561912601b841d6e129ef2e3e1ff4eb7ee3fe6309408a68f9cfe046d7310a4615d2ef96e84fc73249ddcda594cb680b33f5e91cd98d76fc8e3daf930b40c79c6b93a978a2d1fc00b84166e53d0554392be9ad2150b3ec49549795026754241f092ffb9fd547f4af7c3a00b9f92a536c6c05505f0ef6799aacebb3d1fbe333134121ef6298bca7bd292701\",\"mixHash\":\"0x63746963616c2062797a616e74696e65206661756c7420746f6c6572616e6365\",\"nonce\":\"0x0000000000000000\",\"hash\":\"0x61ae2cf19cac775b0297047cb1f75f55158d577a03602fbc1e06a544a2e782b4\",\"transactions\":[],\"uncles\":[]}}"
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.025Z [DEBUG] polygon.consensus.syncer: enqueue block: peer=16Uiu2HAkzM6s7qJvkAzp8yug6D984Q6F2VYCnoUVFin99A7x5vxG number=3098007 hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.026Z [DEBUG] polygon.consensus.syncer: update peer status: peer=16Uiu2HAkzM6s7qJvkAzp8yug6D984Q6F2VYCnoUVFin99A7x5vxG latest block number=3098007 latest block hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2 difficulty=4798825235029
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.027Z [DEBUG] polygon.consensus.syncer: enqueue block: peer=16Uiu2HAmVWJ7TDHtyoJtSXuDgS8mKvUwWWxzxjTsTTFRU31D6xNA number=3098007 hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.028Z [DEBUG] polygon.consensus.syncer: update peer status: peer=16Uiu2HAmVWJ7TDHtyoJtSXuDgS8mKvUwWWxzxjTsTTFRU31D6xNA latest block number=3098007 latest block hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2 difficulty=4798825235029
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.032Z [DEBUG] polygon.consensus.syncer: enqueue block: peer=16Uiu2HAm6yfKYWJncEh8SBtt1ptsavc4M1fySpc8ZKDm8wE9hp3x number=3098007 hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.032Z [DEBUG] polygon.consensus.syncer: update peer status: peer=16Uiu2HAm6yfKYWJncEh8SBtt1ptsavc4M1fySpc8ZKDm8wE9hp3x latest block number=3098007 latest block hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2 difficulty=4798825235029
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.039Z [DEBUG] polygon.consensus.syncer: enqueue block: peer=16Uiu2HAm914jvbG3x2xedF6mEEjqvJj5ZZP9WN2c1R73KH4zJcBL number=3098007 hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.039Z [DEBUG] polygon.consensus.syncer: update peer status: peer=16Uiu2HAm914jvbG3x2xedF6mEEjqvJj5ZZP9WN2c1R73KH4zJcBL latest block number=3098007 latest block hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2 difficulty=4798825235029
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.058Z [DEBUG] polygon.consensus.syncer: enqueue block: peer=16Uiu2HAmVMCHTqD6Xe1KescA6yt8raf13TAgxiWYKRQV3k1jZR7z number=3098007 hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.058Z [DEBUG] polygon.consensus.syncer: update peer status: peer=16Uiu2HAmVMCHTqD6Xe1KescA6yt8raf13TAgxiWYKRQV3k1jZR7z latest block number=3098007 latest block hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2 difficulty=4798825235029
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.060Z [DEBUG] polygon.consensus.syncer: enqueue block: peer=16Uiu2HAm8fbLe2utrGKrBZ3mArgQRAyuAZXY9C6FV3tL3UtnEVBw number=3098007 hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.060Z [DEBUG] polygon.consensus.syncer: update peer status: peer=16Uiu2HAm8fbLe2utrGKrBZ3mArgQRAyuAZXY9C6FV3tL3UtnEVBw latest block number=3098007 latest block hash=0x6521c6f783392db92101a47c477d75e2d2ecb9c26a5358c30f7618e084c397d2 difficulty=4798825235029
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.605Z [DEBUG] polygon.jsonrpc: handle: request="{\"method\":\"eth_getBlockByNumber\",\"params\":[\"latest\",false],\"id\":174,\"jsonrpc\":\"2.0\"}"
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.605Z [DEBUG] polygon.dispatcher: request: method=eth_getBlockByNumber id=174
Apr 20 18:28:26 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:26.605Z [DEBUG] polygon.jsonrpc: handle: response="{\"jsonrpc\":\"2.0\",\"id\":174,\"result\":{\"parentHash\":\"0xc27f5204d57432915729404a3446d91c439ec3156b509403f576cefd44874e19\",\"sha3Uncles\":\"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347\",\"miner\":\"0x0000000000000000000000000000000000000000\",\"stateRoot\":\"0x037444c8f8dc64d3aa0e84979a4452e72908b34cc17dec99c3b7020b95ac92c4\",\"transactionsRoot\":\"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421\",\"receiptsRoot\":\"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421\",\"logsBloom\":\"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\",\"difficulty\":\"0x2f4593\",\"totalDifficulty\":\"0x2f4593\",\"size\":\"0x3a1\",\"number\":\"0x2f4593\",\"gasLimit\":\"0x3938700\",\"gasUsed\":\"0x0\",\"timestamp\":\"0x626050c2\",\"extraData\":\"0x0000000000000000000000000000000000000000000000000000000000000000f90179f86994a3a585c6d59cce6aae7035e8df48b3327cc8be54947eff16a34e3433182d636488bc97919b10283f37943e64f88c6c7a1310236b242180c0ba1409d10f4d949cde4f11ea57163023a9a8a5df499b27813639eb946c230a5f8dd3413aa04e0860dd225f887b36f101b8416f7b503eb78cc23f7509e2ce456726470a58feb3cd7dd38e4e2c12e57cebf710064d3e7f49eb3ac0e92a4f579d3ecc2461986cc5c95025b455939079da72008501f8c9b841e7ff6a9203233b644d863ce2d86dfdefa77ad738d18d22a2cb25167c18384e4358e01a4f1d5f6d9e4b396e65a3f8b47a43444acf6b42fdc79c7cd786f561912601b841d6e129ef2e3e1ff4eb7ee3fe6309408a68f9cfe046d7310a4615d2ef96e84fc73249ddcda594cb680b33f5e91cd98d76fc8e3daf930b40c79c6b93a978a2d1fc00b84166e53d0554392be9ad2150b3ec49549795026754241f092ffb9fd547f4af7c3a00b9f92a536c6c05505f0ef6799aacebb3d1fbe333134121ef6298bca7bd292701\",\"mixHash\":\"0x63746963616c2062797a616e74696e65206661756c7420746f6c6572616e6365\",\"nonce\":\"0x0000000000000000\",\"hash\":\"0x61ae2cf19cac775b0297047cb1f75f55158d577a03602fbc1e06a544a2e782b4\",\"transactions\":[],\"uncles\":[]}}"
Apr 20 18:28:27 ip-172-31-41-6.ec2.internal web[3720]: 2022-04-20T18:28:27.355Z [DEBUG] polygon.jsonrpc: handle: request="{\"method\":\"eth_chainId\",\"params\":[],\"id\":177,\"jsonrpc\":\"2.0\"}"
```

### Proposed solution

Perhaps a flag can be introduced to minimize the delay between syncing blocks from a peer vs actually writing them - either the queue size being used by enqueue or some other kind of logic that forces each block synced to be immediately written.

### PR DESCRIPTION
**Pull request by zivkovicmilos**
# Description

This PR makes the flaky and insane legacy syncer implementation a little more readable.
The only reason it's not `rm -rf`'d is because we're providing backwards compatibility support, otherwise this legacy logic needs to be rewritten completely.

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

Tested manually with older versions of the client.

# Additional comments

Because the [original issue](https://github.com/0xPolygon/polygon-edge/issues/510) that started this PR was resolved by simply notifying peers of new blocks [asynchronously](https://github.com/0xPolygon/polygon-edge/blob/d524ed58dfd09e2281a8c084ab5a5f77bc77719e/protocol/syncer.go#L175) , much of the existing tests in `syncer_test.go` started failing due to not waiting for this dispatch to execute. To alleviate this until syncer_v2 is implemented, a test helper [SyncBroadcast](https://github.com/0xPolygon/polygon-edge/blob/d524ed58dfd09e2281a8c084ab5a5f77bc77719e/protocol/syncer_test.go#L20-L46) was introduced with the same behaviour as the original `Broadcast` - synchronously waiting for peers to be notified.  

Fixes EDGE-557
Fixes EDGE-564
Fixes EDGE-563
Fixes EDGE-561
Fixes EDGE-565

### CODE REVIEW

**Comment by dankostiuk** [comment]
Hey @zivkovicmilos , @dbrajovic ,

After pulling in `bugfix/slow-syncer` into our testnet, I was no longer able to reproduce the issue I opened in https://github.com/0xPolygon/polygon-edge/issues/510. 

I re-tested after pulling in `bugfix/bulk-sync` as well and everything seems fine (I wasn't able to reproduce the issue above). Note that I haven't tested on _just_ this branch (`bugfix/bulk-sync`) but seeing as this refactor looks to include the original changes `bugfix/slow-syncer`, I would expect the same results.

Nice work!

**Comment by dankostiuk** [comment]
Hey @zivkovicmilos , @dbrajovic ,

After pulling in `bugfix/slow-syncer` into our testnet, I was no longer able to reproduce the issue I opened in https://github.com/0xPolygon/polygon-edge/issues/510. 

I re-tested after pulling in `bugfix/bulk-sync` as well and everything seems fine (I wasn't able to reproduce the issue above). Note that I haven't tested on _just_ this branch (`bugfix/bulk-sync`) but seeing as this refactor looks to include the original changes `bugfix/slow-syncer`, I would expect the same results.

Nice work!

File: protocol/syncer.go [hunk 1]
```diff
@@ -596,7 +369,7 @@ func (s *Syncer) WatchSyncWithPeer(p *SyncPeer, handler func(b *types.Block) boo
 			break
 		}
 
-		b, err := p.popBlock(popTimeout)
+		b, err := p.popBlock(blockTimeout * 3)
```
Comment (0xAleksaOpacic) [inline protocol/syncer.go:376]: Why blockTimeout * 3?

File: protocol/syncer.go [hunk 2]
```diff
@@ -609,7 +382,11 @@ func (s *Syncer) WatchSyncWithPeer(p *SyncPeer, handler func(b *types.Block) boo
 			break
 		}
 
-		if handler(b) {
+		exit := newBlockHandler(b)
```
Comment (0xAleksaOpacic) [inline protocol/syncer.go]: i would rename this to shouldExit, because it is bool. But not necessary

File: protocol/syncer.go [hunk 3]
```diff
@@ -700,34 +483,20 @@ func (s *Syncer) BulkSyncWithPeer(p *SyncPeer, newBlockHandler func(block *types
...
-	}
-
-	if len(resp.Objs) != 1 {
-		return nil, fmt.Errorf("unexpected more than 1 result")
-	}
+		pruned := syncPeer.purgeBlocks(block.Hash())
 
-	header := &types.Header{}
+		s.logger.Debug(
+			"pruned peer enqueued blocks",
```
Comment (0xAleksaOpacic) [inline protocol/syncer.go]: Why "blocks" when we are pruning only 1 block

File: protocol/skeleton.go [hunk 4]
```diff
@@ -29,105 +30,67 @@ func getHeaders(clt proto.V1Client, req *proto.GetHeadersRequest) ([]*types.Head
...
-	bodyIndex := []int{}
-
-	for indx, h := range resp {
-		if h.TxRoot != types.EmptyRootHash {
-			bodyHashes = append(bodyHashes, h.Hash)
-			bodyIndex = append(bodyIndex, indx)
+	// Make sure the number sequences match up
+	for i := 1; i < len(headers); i++ {
+		if headers[i].Number-headers[i-1].Number != 1 {
+			return errors.New("invalid header sequence")
```
Comment (0xAleksaOpacic) [inline protocol/skeleton.go]: Why is this not a standardized package error?

File: protocol/skeleton.go [hunk 5]
```diff
@@ -29,105 +30,67 @@ func getHeaders(clt proto.V1Client, req *proto.GetHeadersRequest) ([]*types.Head
...
+	// Grab the block bodies
+	bodies, err := getBodies(getBodiesContext, peerClient, headerHashes)
 	if err != nil {
 		return err
 	}
 
-	for indx, body := range bodies {
-		slot.blocks[bodyIndex[indx]].Transactions = body.Transactions
+	if len(bodies) != len(headers) {
+		return errors.New("requested body and header mismatch")
```
Comment (0xAleksaOpacic) [inline protocol/skeleton.go]: Why is this not a standardized package error?

File: consensus/ibft/ibft.go [hunk 6]
```diff
@@ -541,7 +541,7 @@ func (i *Ibft) runSyncState() {
 			isValidator = i.isValidSnapshot()
 
 			return isValidator
-		})
+		}, i.blockTime)
```
Comment (Kourin1996) [inline consensus/ibft/ibft.go:545]: If `--block-time 0` is given, there will be no timeout?

**Comment by 0xAleksaOpacic** [comment]
https://github.com/0xPolygon/polygon-edge/blob/7d6cd800c75e489ecb02f0b8d480215eae5606b6/protocol/syncer.go#L114 Personally like more 
if !ok{
    //error or print
}
//do stuff

File: protocol/skeleton.go [hunk 7]
```diff
@@ -14,7 +15,7 @@ func getHeaders(clt proto.V1Client, req *proto.GetHeadersRequest) ([]*types.Head
 		return nil, err
 	}
 
-	headers := []*types.Header{}
+	headers := make([]*types.Header, 0)
```
Comment (Kourin1996) [inline protocol/skeleton.go]: If you know the size, please reserve it

File: protocol/skeleton.go [hunk 8]
```diff
@@ -29,105 +30,67 @@ func getHeaders(clt proto.V1Client, req *proto.GetHeadersRequest) ([]*types.Head
...
+	// Construct the body request
+	headerHashes := make([]types.Hash, len(headers))
+	for index, header := range headers {
+		headerHashes[index] = header.Hash
 	}
 
-	bodies, err := getBodies(context.Background(), clt, bodyHashes)
+	getBodiesContext, cancelFn := context.WithTimeout(
+		context.Background(),
+		time.Second*10,
```
Comment (Kourin1996) [inline protocol/skeleton.go]: Please define constant

File: protocol/sync_peer.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+	diff, ok := new(big.Int).SetString(p.Difficulty, 10)
+	if !ok {
+		return nil, fmt.Errorf("failed to decode difficulty")
+	}
+
+	s.Difficulty = diff
+
+	return s, nil
+}
```
Comment (Kourin1996) [inline protocol/sync_peer.go]: What is the difference between these 2 funcs?

File: protocol/sync_peer.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+// from the SyncPeer
+func (s *SyncPeer) purgeBlocks(lastSeen types.Hash) uint64 {
+	s.enqueueLock.Lock()
+	defer s.enqueueLock.Unlock()
+
+	indx := -1
+
+	for i, b := range s.enqueue {
+		if b.Hash() == lastSeen {
+			indx = i
```
Comment (Kourin1996) [inline protocol/sync_peer.go:99]: we can break here?

File: protocol/syncer.go [hunk 11]
```diff
@@ -508,82 +355,8 @@ func (s *Syncer) DeletePeer(peerID peer.ID) error {
...
-		return nil, nil, ErrForkNotFound
-	}
-
-	return header, fork, nil
-}
-
 // WatchSyncWithPeer subscribes and adds peer's latest block
-func (s *Syncer) WatchSyncWithPeer(p *SyncPeer, handler func(b *types.Block) bool) {
+func (s *Syncer) WatchSyncWithPeer(p *SyncPeer, newBlockHandler func(b *types.Block) bool, blockTimeout time.Duration) {
 	// purge from the cache of broadcasted blocks all the ones we have written so far
```
Comment (0xAleksaOpacic) [inline protocol/syncer.go:364]: nitpick: broadcasted is not standard English, "broadcast" is used for past tense :)

File: protocol/syncer.go (hunk 1 above)
Comment (Kourin1996) [inline protocol/syncer.go:376]: 3 is meaning value? Have you tried to test something?
And you need to define it as constant

**Review by Kourin1996 — CHANGES_REQUESTED** [review]

**Review by 0xAleksaOpacic — APPROVED** [review]
Great job with this refactor 🚀 Tested manually also backward compatibility, and everything works.

File: consensus/ibft/ibft.go (hunk 6 above)
Comment (dbrajovic) [reply consensus/ibft/ibft.go:545]: Not something we ever considered to be set to 0. (afaik, if it is, other parts of the system start acting weird)

Do you think there should be an exception given a 0 block time ?

File: protocol/syncer.go (hunk 1 above)
Comment (dbrajovic) [reply protocol/syncer.go:376]: Previously, the `popTimeout` value was hardcoded to `2` without it ever taking into account the configured block time. Factor of `X * 3` is what we think to be a safe estimate of how long you should wait on a block from a peer given some configured block time of value `X`.

File: consensus/ibft/ibft.go (hunk 6 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go:545]: I tried to give 0 to `--block-time`. Then IBFT has 0 in blockTime. After that, syncer does not check this value

File: protocol/syncer.go (hunk 11 above)
Comment (zivkovicmilos) [reply protocol/syncer.go:364]: I'm dialing Pedja as we speak 📞 

This is actually not true, both forms are acceptable:
https://www.merriam-webster.com/dictionary/broadcast

Also:
https://www.merriam-webster.com/words-at-play/is-it-broadcast-or-broadcasted-forecast-or-forecasted-usage

File: protocol/sync_peer.go (hunk 10 above)
Comment (zivkovicmilos) [reply protocol/sync_peer.go:99]: @dbrajovic ?

It looks like we can, but this insane method assumes that the order is guaranteed....

File: protocol/syncer.go (hunk 1 above)
Comment (zivkovicmilos) [reply protocol/syncer.go:376]: Please see:
https://github.com/0xPolygon/polygon-edge/pull/548#discussion_r885519736

cc @dbrajovic

File: protocol/syncer.go (hunk 3 above)
Comment (zivkovicmilos) [reply protocol/syncer.go]: Fixed 👍 

[acbc282](https://github.com/0xPolygon/polygon-edge/pull/548/commits/acbc282cff8e0d593e46b8e89f5623af8164b3f8)

File: protocol/skeleton.go (hunk 4 above)
Comment (zivkovicmilos) [reply protocol/skeleton.go]: Fixed 👍 

[0e4c69b](https://github.com/0xPolygon/polygon-edge/pull/548/commits/0e4c69b7e4de64ad295640a9a83c65cb65bd28e5)

File: protocol/skeleton.go (hunk 5 above)
Comment (zivkovicmilos) [reply protocol/skeleton.go]: Fixed 👍 

[0e4c69b](https://github.com/0xPolygon/polygon-edge/pull/548/commits/0e4c69b7e4de64ad295640a9a83c65cb65bd28e5)

File: consensus/ibft/ibft.go (hunk 6 above)
Comment (zivkovicmilos) [reply consensus/ibft/ibft.go:545]: This is a good catch, not really related to the syncer, as this check should've been present in the actual command 🤦 

Added it 👍 

[733f659](https://github.com/0xPolygon/polygon-edge/pull/548/commits/733f65954c6bbf57c2025b86c0302e7427fd05e1)

File: protocol/skeleton.go (hunk 7 above)
Comment (zivkovicmilos) [reply protocol/skeleton.go]: Reserved 👍 

[5d0f6c7](https://github.com/0xPolygon/polygon-edge/pull/548/commits/5d0f6c7f6d66a7d1a0e779144d350b0a0d464eec)

File: protocol/skeleton.go (hunk 8 above)
Comment (zivkovicmilos) [reply protocol/skeleton.go]: Defined 👍 

[354739f](https://github.com/0xPolygon/polygon-edge/pull/548/commits/354739ff8a5b1e7e69b53ab277c08fb8f5e0dfe4)

File: protocol/syncer.go (hunk 2 above)
Comment (zivkovicmilos) [reply protocol/syncer.go]: @dbrajovic this has got to be one of the nittiest of nitpicks 😆 

Wish granted 👍 

[1924f35](https://github.com/0xPolygon/polygon-edge/pull/548/commits/1924f35ded647462f217e27b6bd9e43c0c20153c)

File: protocol/sync_peer.go (hunk 9 above)
Comment (zivkovicmilos) [reply protocol/sync_peer.go]: You're actually right - there are practically no differences.

Fixed 👍 

[26bb546](https://github.com/0xPolygon/polygon-edge/pull/548/commits/26bb5460fd77e97362f4ec5221de80219e45e09e)

File: protocol/syncer.go (hunk 11 above)
Comment (0xAleksaOpacic) [reply protocol/syncer.go:364]: Lol hahahaha

File: protocol/sync_peer.go (hunk 10 above)
Comment (dbrajovic) [reply protocol/sync_peer.go:99]: Yes, we can.

[Added break](https://github.com/0xPolygon/polygon-edge/pull/548/commits/0ce87e0168ad5964c3aa97bc11459677a7aa7a9e)

**Review by Kourin1996 — APPROVED** [review]

**Pull request merged**